### PR TITLE
C#: Mention more XSS sanitisation options in query help.

### PR DESCRIPTION
**Pull request by max-schaefer**
/cc @erik-krogh

### CODE REVIEW

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp [hunk 1]
```diff
@@ -11,17 +11,33 @@ without properly sanitizing the input first, allows for a cross-site scripting v
...
+The following example shows the page parameter being written directly to the server error page,
+leaving the website vulnerable to cross-site scripting.
+</p>
 <sample src="XSS.cs" />
+<p>
+Sanitizing the user-controlled data using <code>WebUtility.HtmlEncode</code> method prevents the vulnerability:
+</p>
+<sample src="XSSGood.cs" />
+<p>
+Alternatively, the AntiXSS library can be used to sanitize the user-controlled data:
```
Comment (erik-krogh) [inline csharp/ql/src/Security Features/CWE-079/XSS.qhelp:34]:
```suggestion
Alternatively, the <code>AntiXSS</code> NuGet package can be used to sanitize the user-controlled data:
```

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp [hunk 2]
```diff
@@ -11,17 +11,33 @@ without properly sanitizing the input first, allows for a cross-site scripting v
...
+</p>
+<sample src="XSSGood2.cs" />
+<p>
+Recall that this solution requires the AntiXSS library to be installed, for example by
+adding a package reference to the AntiXSS NuGet package to the project file.
+</p>
```
Comment (erik-krogh) [inline csharp/ql/src/Security Features/CWE-079/XSS.qhelp:40]:
```suggestion
<sample src="XSSGood2.cs" />
```

I don't think this last part is needed if we explicitly call it the `AntiXSS` NuGet package above. 
I think developers know that they should add NuGet packages to use them.

**Review by erik-krogh — COMMENTED** [review]
optional comments

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp [hunk 3]
```diff
@@ -11,17 +11,33 @@ without properly sanitizing the input first, allows for a cross-site scripting v
...
 
-<p>To guard against cross-site scripting, consider using contextual output encoding/escaping before
-writing user input to the page, or one of the other solutions that are mentioned in the
-references.</p>
+<p>
+To guard against cross-site scripting, consider using a library providing suitable encoding
+functionality, such as the <code>System.Net.WebUtility</code> class or the AntiXSS library,
+to sanitize the untrusted input before writing it to the page.
+The references also mention other possible solutions.
+</p>
```
[the remark this answers is not in the document]
Comment (max-schaefer) [reply csharp/ql/src/Security Features/CWE-079/XSS.qhelp:19]: I'm not sure about this suggestion. Sentences starting with conjunctions don't read well.

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp [hunk 4]
```diff
@@ -11,17 +11,29 @@ without properly sanitizing the input first, allows for a cross-site scripting v
 </overview>
 <recommendation>
 
-<p>To guard against cross-site scripting, consider using contextual output encoding/escaping before
-writing user input to the page, or one of the other solutions that are mentioned in the
-references.</p>
+<p>
+To guard against cross-site scripting, consider using a library providing suitable encoding
+functionality, such as the <code>System.Net.WebUtility</code> class or the <code>AntiXSS</code> NuGet package,
```
Comment (tamasvajk) [inline csharp/ql/src/Security Features/CWE-079/XSS.qhelp:16]: I didn't know this `AntiXSS` package. It doesn't seem to be maintained. Its last version was released on 6/2/2014, and its project website doesn't exist any longer: https://wpl.codeplex.com/. I don't think we should mention it.

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp [hunk 5]
```diff
@@ -11,17 +11,29 @@ without properly sanitizing the input first, allows for a cross-site scripting v
...
 <example>
 
-<p>The following example shows the page parameter being written directly to the server error page,
-leaving the website vulnerable to cross-site scripting.</p>
-
+<p>
+The following example shows the page parameter being written directly to the server error page,
+leaving the website vulnerable to cross-site scripting.
+</p>
 <sample src="XSS.cs" />
```
Comment (tamasvajk) [inline csharp/ql/src/Security Features/CWE-079/XSS.qhelp:28]: I would rename this file to `XSSBad.cs` to make it more explicit.

**Comment by max-schaefer** [comment]
Thanks for your review, @tamasvajk! Comments addressed.

**Review by tamasvajk — APPROVED** [review]

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp [hunk 6]
```diff
@@ -11,17 +11,24 @@ without properly sanitizing the input first, allows for a cross-site scripting v
 </overview>
 <recommendation>
 
-<p>To guard against cross-site scripting, consider using contextual output encoding/escaping before
-writing user input to the page, or one of the other solutions that are mentioned in the
-references.</p>
+<p>
+To guard against cross-site scripting, consider using a library providing suitable encoding
+functionality, such as the <code>System.Net.WebUtility</code> class to sanitize the untrusted input before writing it to the page.
```
Comment (felicitymay) [inline csharp/ql/src/Security Features/CWE-079/XSS.qhelp:16]: Suggest a minor change:

```suggestion
To guard against cross-site scripting, consider using a library that provides suitable encoding
functionality, such as the <code>System.Net.WebUtility</code> class, to sanitize the untrusted input before writing it to the page.
```

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp (hunk 3 above)
[the remark this answers is not in the document]
Comment (felicitymay) [reply csharp/ql/src/Security Features/CWE-079/XSS.qhelp:18]: Maybe:

```suggestion
For other possible solutions, see the references.
</p>
```

File: csharp/ql/src/Security Features/CWE-079/XSS.qhelp [hunk 7]
```diff
@@ -11,17 +11,24 @@ without properly sanitizing the input first, allows for a cross-site scripting v
...
-leaving the website vulnerable to cross-site scripting.</p>
-
-<sample src="XSS.cs" />
+<p>
+The following example shows the page parameter being written directly to the server error page,
+leaving the website vulnerable to cross-site scripting.
+</p>
+<sample src="XSSBad.cs" />
+<p>
+Sanitizing the user-controlled data using <code>WebUtility.HtmlEncode</code> method prevents the vulnerability:
```
Comment (felicitymay) [inline csharp/ql/src/Security Features/CWE-079/XSS.qhelp:29]:
```suggestion
Sanitizing the user-controlled data using the <code>WebUtility.HtmlEncode</code> method prevents the vulnerability:
```

**Review by felicitymay — DISMISSED** [review]
Thanks for the improvements to the query help. I made a couple of small suggestions, but it's great to see a `good` example.

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**